projects/components/src/download-file/service/file-download.service.ts

```diff
@@ -0,0 +1,130 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { Dictionary } from '@hypertrace/common';
+import { isEmpty, startCase } from 'lodash-es';
+import { BehaviorSubject, combineLatest, Observable, of, Subscription } from 'rxjs';
+import { catchError, map, take } from 'rxjs/operators';
+import { NotificationService } from '../../notification/notification.service';
+
+@Injectable()
+export class FileDownloadService implements OnDestroy {
+  private readonly downloadElement: HTMLAnchorElement;
+  private readonly subscriptions: Subscription = new Subscription();
+  private readonly fileDownloadEventSubject: BehaviorSubject<FileDownloadEventType | undefined> = new BehaviorSubject<
+    FileDownloadEventType | undefined
+  >(undefined);
+
+  public readonly fileDownloadEvent$: Observable<
+    FileDownloadEventType | undefined
+  > = this.fileDownloadEventSubject.asObservable();
+
+  public constructor(private readonly notificationService: NotificationService) {
+    this.downloadElement = this.createDownloadElement();
+  }
+
+  /**
+   * Downloads data as text content
+   * @param config Text download config
+   */
+  public downloadAsText(config: FileDownloadBaseConfig): void {
+    this.download({ ...config }, data => `data:text/plain;charset=utf-8,${encodeURIComponent(data)}`);
+  }
+
+  /**
+   * Downloads data as csv formatted content
+   * @param config Csv download config
+   */
+  public downloadAsCsv(config: CsvDownloadFileConfig): void {
+    // If given header is empty, then create header from the data keys
+    const header$ = isEmpty(config.header)
+      ? config.dataSource.pipe(
+          map(data => Object.keys(data[0] ?? [])),
+          map(keys => keys.map(startCase))
+        )
+      : of(config.header!);
+
+    // Value replacer for null and undefined values
+    const replacer = (_: string, value: string) => value ?? '-';
```

Comment:
So if you see at L50, where we're using this function. The callback has two attributes (key, value), since we do not need a key here, I used underscore since it doesn't give a `not used` error.
It is actually a pattern that we use. We could have used `_key` as well